src/Umbraco.Examine/ContentIndexPopulator.cs

```diff
@@ -70,31 +77,89 @@ protected override void PopulateIndexes(IReadOnlyList<IIndex> indexes)
             {
                 contentParentId = _parentId.Value;
             }
+
+            if (_publishedValuesOnly)
+            {
+                IndexPublishedContent(contentParentId, pageIndex, pageSize, indexes);
+            }
+            else
+            {
+                IndexAllContent(contentParentId, pageIndex, pageSize, indexes);
+            }
+        }
+
+        protected void IndexAllContent(int contentParentId, int pageIndex, int pageSize, IReadOnlyList<IIndex> indexes)
+        {
             IContent[] content;
 
             do
             {
-                if (!_publishedValuesOnly)
-                {
-                    content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _).ToArray();
-                }
-                else
+                content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _).ToArray();
+
+                if (content.Length > 0)
                 {
-                    //add the published filter
-                    //note: We will filter for published variants in the validator
-                    content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _,
-                        _publishedQuery, Ordering.By("Path", Direction.Ascending)).ToArray();
+                    // ReSharper disable once PossibleMultipleEnumeration
+                    foreach (var index in indexes)
+                        index.IndexItems(_contentValueSetBuilder.GetValueSets(content));
                 }
 
+                pageIndex++;
+            } while (content.Length == pageSize);
+        }
+
+        protected void IndexPublishedContent(int contentParentId, int pageIndex, int pageSize,
+            IReadOnlyList<IIndex> indexes)
+        {
+            IContent[] content;
+
+            var publishedPages = new HashSet<int>();
+
+            do
+            {
+                //add the published filter
+                //note: We will filter for published variants in the validator
+                content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _, _publishedQuery,
+                    Ordering.By("Level", Direction.Ascending)).ToArray();
+               
                 if (content.Length > 0)
                 {
+                    var indexableContent = new List<IContent>();
+
+                    // get the max level in this result set
+                    int maxLevel = content.Max(x => x.Level);
+
+                    // gets the first level pages, these are always published because _publishedQuery filter
+                    // and store the id in a hash set so we can track them
+                    var firstLevelContent = content.Where(x => x.Level == 1);
```

Comment:
Why do we need to do this? This means that we are doing to iterate over the entire result set more than once which we don't want to do. This also means we are going to iterate over all subsequent paged results sets even though we would know we no longer have any level 1s remaining. We already know we are ordering by level so we know the first content is going to be level one which can be tracked. We should be able to achieve what is needed without iterating the result set more than once. We manage this this way in v7 https://github.com/umbraco/Umbraco-CMS/blob/v7/dev/src/UmbracoExamine/UmbracoContentIndexer.cs#L497

Reply:
I did this check because we are in a loop. If somebody more than 10000 root level nodes we need to keep track of this somehow. But I will see if I can optimize

Reply:
Refactored to everything in a single foreach